File: BEmuJava/bemu/src/main/java/com/bemu/BEmu/Name.java
//------------------------------------------------------------------------------
// <copyright project="BEmuJava" file="BEmu.Name.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bemu.BEmu;
import java.util.HashMap;
import java.util.Map;

public class Name
{	
	//private final boolean _isNull;
	private final String _name;
	private final static Map<String, Name> _globalNameTable = new HashMap<String, Name>();
	
	public static final Name NullName = new Name();
	
	public Name()
	{
	    this._name = null;
	    //this._isNull = true;
	}
	
	public Name(String nameString) throws Exception
	{
		if(nameString == null)
			throw new Exception("nameString can't be null");
		
		this._name = nameString;
		//this._isNull = false;
		
		if(!Name._globalNameTable.containsKey(nameString))
			Name._globalNameTable.put(nameString, this);
	}
	
	public static Name findName(String nameString)
	{
	    if(Name._globalNameTable.containsKey(nameString))
	    	return Name._globalNameTable.get(nameString);
	    else
	    	return null;
	}
	
	public static Name getName(String nameString) throws Exception
	{
	    if(Name._globalNameTable.containsKey(nameString))
	    	return Name._globalNameTable.get(nameString);
	    else
	    	return new Name(nameString);
	}
	
    public static boolean hasName(String nameString)
    {
        return Name._globalNameTable.containsKey(nameString);
    }
    
	public String toString()
	{
		return this._name;
	}
	
	public boolean equals(Object obj)
	{
		if(obj.getClass() == this.getClass())
			return ((Name)obj)._name.equals(this._name);
		else if(obj.getClass() == String.class)
			return this._name.equals((String)obj);
		else
			return false;
	}
}

File: BEmu_cpp/src/BloombergTypes/Name.cpp
//------------------------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/BloombergTypes/Name.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------------------------

#include "BloombergTypes/Name.h"
#include "BloombergTypes/ElementPtr.h"

namespace BEmu
{
	std::map<std::string, Name> Name::_globalNameTable;

	Name::Name()
	{
		this->_isNull = true;
	}

	Name::Name(const char* nameString)
	{
		this->_isNull = false;
		this->_name = nameString;
		this->_cname = nameString;
		this->_length = strlen(nameString);

		if(Name::_globalNameTable.find(this->_name) == Name::_globalNameTable.end()) //if name doesn't already exist
			Name::_globalNameTable[this->_name] = *this;
	}

	Name::Name(const Name& original)
	{
		this->_isNull = original._isNull;
		this->_name = original._name;
		this->_cname = this->_name.c_str();
		this->_length = original._length;
	}

	Name& Name::operator=(const Name &rhs)
	{
		if (this != &rhs)
		{
			this->_isNull = rhs._isNull;
			this->_name = rhs._name;
			this->_cname = this->_name.c_str();
			this->_length = rhs._length;
		}
		return *this;
	}

	void Name::clearGlobalNameTable()
	{
		Name::_globalNameTable.clear();
	}

	size_t Name::length() const
	{
		size_t result = this->_name.length();
		return result;
	}

	size_t Name::hash() const
	{
		if(this->_isNull)
			return 0;
		else
		{
			int hash = 17;
			for(size_t i = 0; i < this->_name.length(); i++)
			{
				hash = hash * 23 + this->_name[i];
			}

			return hash;
		}
	}

	Name Name::findName(const char *nameString)
	{
		if(Name::_globalNameTable.find(nameString) == Name::_globalNameTable.end()) //if name doesn't exist
		{
			Name result(nameString); //create the name and insert it into _globalNameTable
			return result;
		}
		else //if name already exists
			return Name::_globalNameTable[nameString];
	}

	bool Name::hasName(const char *nameString)
	{
		return Name::_globalNameTable.find(nameString) != Name::_globalNameTable.end();
	}

	bool Name::isNull() const
	{
		return this->_isNull;
	}


	const char* Name::string() const
	{
		return this->_cname;
	}

	std::ostream& operator<<(std::ostream& stream, const Name& name)
	{
		return stream << name.string();
	}

	bool operator==(const Name& lhs, const char *rhs)
	{
		bool lhsIsNull = lhs.isNull();
		bool rhsIsNull = rhs == "";

		if(lhsIsNull)
			return rhsIsNull;
		else if(rhsIsNull)
			return false;
		else
			return strncmp(lhs.string(), rhs, lhs.length() + 1) == 0;
	}

	bool operator==(const Name& lhs, const Name rhs)
	{
		bool lhsIsNull = lhs.isNull();
		bool rhsIsNull = rhs.isNull();

		if(lhsIsNull)
			return rhsIsNull;
		else if(rhsIsNull)
			return false;
		else
			return strncmp(lhs.string(), rhs.string(), lhs.length() + 1) == 0;
	}

	bool operator==(const char *lhs, const Name& rhs)
	{
		return rhs == lhs;
	}

	bool operator!=(const Name& lhs, const char *rhs)
	{
		return !(lhs == rhs);
	}

	bool operator!=(const Name& lhs, const Name rhs)
	{
		return !(lhs == rhs);
	}

	bool operator!=(const char *lhs, const Name& rhs)
	{
		return rhs == lhs;
	}

	//Note that while the order of this operation is consistent, it's not guaranteed to be lexical (this is the same in the BB API)
	//I made null Names less than Names with any value.  I haven't checked this against what the BB API returns.
	bool operator<(const Name& lhs, const Name& rhs)
	{
		bool lhsIsNull = lhs.isNull();
		bool rhsIsNull = rhs.isNull();

		if(lhsIsNull && rhsIsNull)
			return false;
		else if(lhsIsNull)
			return true;
		else if(rhsIsNull)
			return false;
		else
			return lhs.hash() < rhs.hash();
	}

	bool operator<=(const Name& lhs, const Name& rhs)
	{
		return (lhs < rhs) || (lhs == rhs);
	}

	bool operator>(const Name& lhs, const Name& rhs)
	{
		return !(lhs <= rhs);
	}

	bool operator>=(const Name& lhs, const Name& rhs)
	{
		return !(lhs < rhs);
	}

}